File: samples/nodejs/signature.js
'use strict';

const uuid = require('uuid');
const crypto = require('crypto-js');

class Signature {
    constructor(apiKey, apiSecret) {
        if (this.apikey || this.apiSecret) {
            throw new Error('apikey and apiSecret must be suppiled to calculate signature');
        }
        this.apikey = apiKey;
        this.apiSecret = apiSecret;
    }


    calculate(nonce, timestamp){
        if (!nonce){
            nonce = uuid.v4();
        }

        if (!timestamp){
            timestamp = new Date().toUTCString();
        }

        // format raw signature
        const signature = `date: ${timestamp}\nx-mod-nonce: ${nonce}`;
        console.debug(`Raw signature string is:\n-----------\n${signature}\n-----------`);
    
        // sign and encode signature
        const signatureSigned = crypto.HmacSHA1(signature, this.apiSecret);
        const signatureEncoded = encodeURIComponent(crypto.enc.Base64.stringify(signatureSigned));
        console.debug(`Signed signature is [${signatureSigned}] and encoded is [${signatureEncoded}]`);
    
        return new Result(this.apikey, timestamp, nonce, signatureEncoded);
    }

}

class Result {
    constructor(key, timestamp, nonce, encodedSignature){
        this.timestamp = timestamp;
        this.nonce = nonce;
        this.encodedSignature = encodedSignature;
        this.authorisation = `Signature keyId="${key}",algorithm="hmac-sha1",headers="date x-mod-nonce",signature="${encodedSignature}"`;
    }

    getTimeStamp(){
        return this.timestamp;
    }

    getHTTPHeaders(){
        return {
            'Date': this.timestamp,
            'x-mod-nonce': this.nonce,
            'Authorization': this.authorisation
        };
    }

    getSignature(){
        return this.encodedSignature;
    }
}

module.exports = Signature, Result;

File: samples/scala/src/main/scala/Signature.scala
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.{Base64, Calendar, Date, Locale, TimeZone, UUID}

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class Signature(val apiKey: String, val apiSecret: String) {

  private val DatePattern: String = "EEE, dd MMM yyyy HH:mm:ss z"
  private val HmacSha1Algorithm: String = "HmacSHA1"
  private val AutorizationHeader: String = "Authorization"
  private val DateHeader: String = "Date"
  private val NonceHeader: String = "x-mod-nonce"

  def buildHeaders(nonce: String = UUID.randomUUID().toString,
                   date: Date = Calendar.getInstance().getTime): Map[String, String] = {
    val timestamp = formatDate(date)
    val signatureString = s"date: $timestamp\nx-mod-nonce: $nonce"

    val hmacSignature = calculateHmac(signatureString)

    Map(AutorizationHeader -> formatAuthHeader(apiKey, hmacSignature),
      DateHeader -> timestamp,
      NonceHeader -> nonce)
  }

  private def formatAuthHeader(key: String, signature: String): String = {
    s"Signature keyId= ${'"'}$key${'"'}," +
      s"algorithm=${'"'}hmac-sha1${'"'}," +
      s"headers=${'"'}date x-mod-nonce${'"'}," +
      s"signature=${'"'}$signature${'"'}"
  }

  private def calculateHmac(message: String): String = {
    val signingKey = new SecretKeySpec(apiSecret.getBytes, HmacSha1Algorithm)

    val mac = Mac.getInstance(HmacSha1Algorithm)
    mac.init(signingKey)

    val rawHmac = mac.doFinal(message.getBytes())
    val hmac = Base64.getEncoder.encodeToString(rawHmac)
    URLEncoder.encode(hmac, StandardCharsets.UTF_8.name())
  }

  private def formatDate(date: Date): String = {
    val simpleDateFormat = new SimpleDateFormat(DatePattern, Locale.UK)
    simpleDateFormat.setTimeZone(TimeZone.getTimeZone("GMT"))
    simpleDateFormat.format(date)
  }
}